File: src/main/java/com/github/fherbreteau/gatling/sftp/javaapi/protocol/SftpProtocolBuilder.java
package com.github.fherbreteau.gatling.sftp.javaapi.protocol;

import io.gatling.core.protocol.Protocol;
import io.gatling.javaapi.core.ProtocolBuilder;
import org.apache.sshd.client.SshClient;

import java.nio.file.Path;

public class SftpProtocolBuilder implements ProtocolBuilder {

    private final com.github.fherbreteau.gatling.sftp.protocol.SftpProtocolBuilder wrapped;

    public SftpProtocolBuilder(com.github.fherbreteau.gatling.sftp.protocol.SftpProtocolBuilder wrapped) {
        this.wrapped = wrapped;
    }

    public SftpProtocolBuilder client( SshClient client) {
        return new SftpProtocolBuilder(wrapped.client(client));
    }

    public SftpProtocolBuilder server( String server) {
        return new SftpProtocolBuilder(wrapped.server(server));
    }

    public SftpProtocolBuilder port(int port) {
        return new SftpProtocolBuilder(wrapped.port(port));
    }

    public SftpProtocolBuilder credentials(String username, String password) {
        return new SftpProtocolBuilder(wrapped.credentials(username, password));
    }

    public SftpProtocolBuilder localSourcePath(Path sourcePath) {
        return new SftpProtocolBuilder(wrapped.localSourcePath(sourcePath));
    }

    public SftpProtocolBuilder localDestinationPath(Path destpath) {
        return new SftpProtocolBuilder(wrapped.localDestinationPath(destpath));
    }

    public SftpProtocolBuilder remoteSourcePath(Path sourcePath) {
        return new SftpProtocolBuilder(wrapped.remoteSourcePath(sourcePath));
    }

    public SftpProtocolBuilder remoteDestinationPath(Path destpath) {
        return new SftpProtocolBuilder(wrapped.remoteDestinationPath(destpath));
    }

    @Override
    public Protocol protocol() {
        return wrapped.protocol();
    }
}

File: src/main/scala/io/github/fherbreteau/gatling/sftp/protocol/SftpProtocolBuilder.scala
package io.github.fherbreteau.gatling.sftp.protocol

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression
import io.gatling.internal.quicklens._
import io.github.fherbreteau.gatling.sftp.model.Authentications
import io.github.fherbreteau.gatling.sftp.util.SftpHelper
import org.apache.sshd.client.ClientBuilder

import java.nio.file.Path

object SftpProtocolBuilder {

  implicit def toSftpProtocol(builder: SftpProtocolBuilder): SftpProtocol = builder.build

  def apply(configuration: GatlingConfiguration): SftpProtocolBuilder = SftpProtocolBuilder(SftpProtocol(configuration))
}

final case class SftpProtocolBuilder(protocol: SftpProtocol) {
  def client(clientBuilder: ClientBuilder): SftpProtocolBuilder = this.modify(_.protocol.exchange.clientBuilder).setTo(clientBuilder)

  def server(server: String): SftpProtocolBuilder = this.modify(_.protocol.exchange.server).setTo(server)

  def port(port: Int): SftpProtocolBuilder = this.modify(_.protocol.exchange.port).setTo(port)

  def password(username: Expression[String], password: Expression[String]): SftpProtocolBuilder =
    this.modify(_.protocol.exchange.authType).setTo(Authentications.Password)
      .modify(_.protocol.credentials).setTo(SftpHelper.buildPasswordAuth(username, password))

  def keyPair(username: Expression[String], keyPath: Expression[String]): SftpProtocolBuilder =
    this.modify(_.protocol.exchange.authType).setTo(Authentications.KeyPair)
      .modify(_.protocol.credentials).setTo(SftpHelper.buildKeyPairAuth(username, keyPath))

  def keyPair(username: Expression[String], keyPath: Expression[String], keyPassphrase: Expression[String]): SftpProtocolBuilder =
    this.modify(_.protocol.exchange.authType).setTo(Authentications.KeyPair)
      .modify(_.protocol.credentials).setTo(SftpHelper.buildKeyPairAuth(username, keyPath, keyPassphrase))

  def localPath(path: Path): SftpProtocolBuilder = this.localSourcePath(path).localDestinationPath(path)

  def localSourcePath(sourcePath: Path): SftpProtocolBuilder = this.modify(_.protocol.localSourcePath).setTo(Some(sourcePath))

  def localDestinationPath(destPath: Path): SftpProtocolBuilder = this.modify(_.protocol.localDestinationPath).setTo(Some(destPath))

  def remotePath(path: String): SftpProtocolBuilder = this.remoteSourcePath(path).remoteDestinationPath(path)

  def remoteSourcePath(sourcePath: String): SftpProtocolBuilder = this.modify(_.protocol.remoteSourcePath).setTo(Some(sourcePath))

  def remoteDestinationPath(destPath: String): SftpProtocolBuilder = this.modify(_.protocol.remoteDestinationPath).setTo(Some(destPath))

  def build: SftpProtocol = protocol
}